### ISSUE TITLE
Add server side command handler plugins

### ISSUE THREAD

**Issue by jerbia** [docker/docker#14233]
Hi All,

Now that volume plugins were added to Docker experimental channel, I would like to suggest adding server-side plugins for handling Docker commands. This will allow Docker users to achieve similar capabilities as supplied by [PowerStrip](https://github.com/ClusterHQ/powerstrip).

The server side command handler plugins are pre/post extension to Docker command handling, allowing to modify input commands before processed by Docker engine, and output data after processed by Docker Engine (but before sent to client). It will allow adding new capabilities into Docker engine, implemented as out-of-process plugins.

I've implemented a similar mechanism for a project I am working on. I'll be glad to get your feedback on the design and contribute the code if it makes sense.

## Plugin Mechanics

The Command Handler plugins are similar to [Volume Plugins](https://github.com/docker/docker/blob/master/experimental/plugin_api.md): 
* Processes running on same host as Docker Engine. 
* Registered by placing a file in `/usr/share/docker/plugins/srv/cmd`. 
* There can be two type of plugin files:
    * `.sock` files are UNIX domain sockets.
    * `.spec` files are text files containing a URL, such as unix:///other.sock.
* The name of the file (excluding the extension) determines the plugin name.
* Plugins should be started before Docker, and stopped after Docker.
* The Plugin API is RPC-style JSON over HTTP. Requests flow from the Docker daemon to the plugin.

## Plugin Initialization
Unlike volume plugins, the command handler plugins should be initialized upon Docker Engine startup. 
Docker Engine sends a /Plugin.Activate request and receives list of commands and type of handling (input or output) that plugin is interested in.

### /Plugin.Activate

**Request:** empty body

**Response:**
```
{
    "Implements": [
        {"cmd":"run", "type":"input"}, 
        {"cmd":"stop", "type":"input"}, 
        {"cmd":"inspect", "type":"output"}
    ]
}
```

## Plugin Invocation
Once a plugin is registered for set of commands, Docker Engine calls the plugin before processing the command ("input" command handler) and before sending the output to client ("output" command handler). 

Plugin invocation is done by calling /Plugin.Execute request, while passing type of processing (input or output) and the associated payload (complete input command or output message).

### /Plugin.Execute

**Request:** 
```
{
    "Command": "run",
    "Type": "input" or "output"
    "Payload": { COMMAND PAYLOAD }
}
```

**Response:**
```
{
    "action": "continue" or "stop"
    "Payload": {UPDATED PAYLOAD}
    "Error": {OPTIONAL ERROR MESSAGE}        
}
```

Docker Engine should continue processing the command in case the plugin responded with "continue" response action. It should stop processing the command (and returning an error) in case the response action is "stop".

In case if "continue" action, Docker Engine should use the Payload returned from the plugin response instead of the original command payload.
This allows plugins to change command input parameters or output data.

Note that multiple command handler plugins can be registered for the same command. Docker Engine will continue invoking plugins for the same command as long as it did not get "stop" action from one of the handlers. After processing all plugins registered for a command, Docker Engine should continue handling the modified command (as returned from the last plugin).

## Example Plugins
Example for capabilities we would be able to implement using command handler plugins:
* Run additional commands before/after container is started/stopped. E.g. when containerA is started/stopped, automatically start/stop containerB.
* Add missing parameters to command line. E.g. add security parameters (cap-add, cap-drop) even if not specified in command line.
* Prevent running command with certain parameters.
* Modify Docker output to include additional data. E.g. modifying 'docker inspect' output to include container start time (or anything else which is useful). Note that this one is tricky as the CLI might be expecting certain outputs.

**Comment by tgraf** [comment]
+1

Although the proposal may be a little bit too generic. Modification of requests and responses is highly desirable to finally get rid of all the wrappers.

**Comment by kostickm** [comment]
@jerbia There have been talk about adding hooks like you are suggesting here. Right now a few of us are working on splitting the current API code out so it can be used by both Docker and Swarm reducing the redundancy. I want to recommend waiting until the API split work is completed so both Docker and Swarm will be able to leverage these plugin hooks by sharing the API code.

Work in progess PoC: https://github.com/kostickm/docker/tree/api-split-poc

cc: @icecrime @calavera @duglin

**Comment by tgraf** [comment]
@kostickm +10 for split and base of hooks on top of that to make it available for Docker and Swarm. Let us know if we can help in any way.

**Comment by jerbia** [comment]
@kostickm sounds like a great idea! 
Having these hooks work for Swarm will provide so many possibilities!

@tgraf the proposal is indeed generic. it is flexible enough to provide a new stack of implementations: security, accessibility, integration with 3rdparties (like slack and ifttt), etc.

**Comment by aanm** [comment]
+1

**Comment by aanm** [comment]
@jerbia Any news about this?

**Comment by jerbia** [comment]
@aanm - There is work being done with regards to command plugins for authorization purposes:
https://github.com/docker/docker/pull/15365

Some of the use cases mentioned in the 'generic command handler plugin' are redundant with the authorization plugins. Would probably make sense to extend the authorization plugin to include missing use cases, like the ability to modify response data.

**Comment by aanm** [comment]
@jerbia and regarding that part have you start working on any code or thinking beside this proposal?

**Comment by jerbia** [comment]
@aanm yes, work was done beside this proposal to check feasibility and implementation details. My current thinking is to include this work on top of the authorization plug-in so that there will be a common ground for command-handler plugins.

**Comment by aanm** [comment]
@jerbia awesome, do you have a link for that? That way I could start writing a plugin as you go.

**Comment by jerbia** [comment]
@aanm sure. let me stabilize what I have so that it will work on environments other than my laptop.

**Comment by aanm** [comment]
@jerbia ping?

**Comment by aanm** [comment]
@jerbia can you give me any update?
